consensus/src/quorum_store/batch_generator.rs

```diff
@@ -261,31 +244,52 @@ impl BatchGenerator {
+            Duration::from_millis(self.config.back_pressure.increase_duration_ms);
+        let mut back_pressure_decrease_latest = Instant::now();
+        let mut back_pressure_increase_latest = Instant::now();
+        let mut dynamic_max_pull_count = (self.config.back_pressure.dynamic_min_batch_count
+            + self.config.back_pressure.dynamic_max_batch_count)
+            / 2;
 
         loop {
             let _timer = counters::WRAPPER_MAIN_LOOP.start_timer();
 
             tokio::select! {
+                biased;
+                Some(updated_back_pressure) = back_pressure_rx.recv() => {
+                    self.back_pressure = updated_back_pressure;
+                },
                 _ = interval.tick() => {
-                    if self.qs_back_pressure || self.block_store.back_pressure() {
-                        counters::QS_BACKPRESSURE.set(1);
-                        // quorum store needs to be back pressured
-                        // if last txn pull is not back pressured, there may be unfinished batch so we need to end the batch
-                        if !back_pressure_in_last_pull {
-                            if let Some(proof_rx) = self.handle_scheduled_pull(true).await {
-                                proofs_in_progress.push(Box::pin(proof_rx));
-                            }
+                    if self.back_pressure {
+                        // multiplicative decrease, every second
+                        if back_pressure_decrease_latest.elapsed() >= back_pressure_decrease_duration {
+                            back_pressure_decrease_latest = Instant::now();
+                            dynamic_max_pull_count = std::cmp::max(
+                                (dynamic_max_pull_count as f64 * self.config.back_pressure.decrease_fraction) as u64,
+                                self.config.back_pressure.dynamic_min_batch_count,
+                            );
+                            debug!("QS: dynamic_max_pull_count: {}", dynamic_max_pull_count);
                         }
-                        back_pressure_in_last_pull = true;
+                        counters::QS_BACKPRESSURE.set(1);
+                        counters::QS_BACKPRESSURE_DYNAMIC_MAX.set(dynamic_max_pull_count as i64);
                     } else {
-                        counters::QS_BACKPRESSURE.set(0);
-                        // no back pressure
-                        if let Some(proof_rx) = self.handle_scheduled_pull(false).await {
-                            proofs_in_progress.push(Box::pin(proof_rx));
+                        // additive increase, every second
+                        if back_pressure_increase_latest.elapsed() >= back_pressure_increase_duration {
+                            back_pressure_increase_latest = Instant::now();
+                            dynamic_max_pull_count = std::cmp::min(
+                                dynamic_max_pull_count + self.config.back_pressure.dynamic_min_batch_count,
```

Comment:
this seems not symmetric?

Reply:
For additive increase, the max_pull_count is increased linearly by the min_batch_count during each duration. This is similar to at least the first versions of TCP.